### PR TITLE
Final changes for the mfem-4.2 release

### PR DESCRIPTION
**Pull request by tzanio**
### Release Checklist

- [x] Update the MFEM version in the following files:
    - [x] `CHANGELOG`
    - [x] `makefile`
    - [x] `CMakeLists.txt`
    - [x] `doc/CodeDocumentation.conf.in`
- [x] Check that version requirements for each of MFEM's dependencies are documented in `INSTALL` and up-to-date
- [x] Check that continuous integration server configurations reflect the dependency version requirements of the new release
    - [x] `.travis.yml`
    - [x] `.appveyor.yml`
- [x] ~(LLNL only) Make sure all `README.html` files in the source repo are up to date.~
- [x] Remove `README.html` files and web documentation from the repository
- [x] Reorganize the CHANGELOG to make it in a coherent story
- [x] Review the whole source once over to check if anything stands out
- [x] Ask applications  to give the branch a try ✅ @artv3 ✅ @jamiebramwell 
- [x] Test on additional platforms and compilers ✅ Lassen ✅ Quartz ✅ Corona ✅ rzwhamo ✅ PGI ✅ Syrah

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1827](https://github.com/mfem/mfem/pull/1827) | @tzanio | @tzanio | @v-dobrev + @adrienbernede + @acfisher + @jamiebramwell + @artv3 | 10/20/20 | 10/24/20 | 10/30/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
@artv3 and @jamiebramwell -- can you try this branch with your applications and let us know if you are seeing any issues?

We'd like to tag the mfem-4.2 release on Monday (10/26/20)

**Comment by tzanio** [comment]
Question for @mfem/developers (and everyone else) -- Should we keep the `README.html` file in the repository or should we get rid of them as part of this release?

Does anybody use them? Will you miss them if they are gone?

The current policy is that the README.html inside e.g. examples is essentially the same as https://mfem.org/examples/ BUT you don’t need internet connection to see it. 

Back in the day, this was useful but these days maybe it is less relevant...

In the meantime the number of examples and the image sizes have grown, making the source repo larger than it should be, see e.g. https://github.com/mfem/mfem/pull/1827/commits/a78b03f9a0555e743970835de768eb3f592c182b.

What do you think?

**Comment by cjvogl** [comment]
> Does anybody use them? Will you miss them if they are gone?

I am just learning this html exists... which I suppose is my answer.

**Comment by dylan-copeland** [comment]
I second @cjvogl , and I am not concerned about dialing into AOL to view the README.

**Comment by jakubcerveny** [comment]
I won't miss the offline examples either...

**Comment by tzanio** [comment]
Ok, there doesn't seem to be any support for the `README.html` files so we will remove them in this release

**Comment by adrienbernede** [comment]
3 commits were removed from this branch.


:rotating_light: Warning: If you don’t know how to deal with that, or run into unexpected errors, or simply want to be sure, **please contact me**.

I you had this branch cloned somewhere, make sure to update it so that it does not contain those 3 commits anymore:
- Those using the branch read-only should only have to force the pull, or try the method below.
- Those with changes on top of it will have to deal with that. I suggest an interactive rebase to remove those 3 commits

```
* d13e719f - Added libceed image (19 hours ago) <Tzanio>
* abeb127c - Added README.html file for Navier (20 hours ago) <Tzanio>
* a78b03f9 - Updated the README.html files (20 hours ago) <Tzanio>
```

To do so from your local clone on mfem-4.2-dev, you may run:
```
git rebase -i 41b0aa5d
```
and remove the three lines corresponding to those commits (the first 3 lines actually).

**Comment by tzanio** [comment]
FYI, I updated my local `mfem-4.2-dev` branch (which had no changes) with
```
git fetch origin mfem-4.2-dev && git checkout -B mfem-4.2-dev origin/mfem-4.2-dev
```

**Review by artv3 — APPROVED** [review]
Awesome! Excited for MFEM 4.2!

**Review by jamiebramwell — APPROVED** [review]
All good here, modulo this memory error that has already been noticed: https://github.com/mfem/mfem/issues/1750

**Comment by tzanio** [comment]
Thanks @artv3 and @jamiebramwell !

**Comment by tzanio** [comment]
Note: the `branch-history` failures in Travis are due to the removal of the `docs/web` directory in https://github.com/mfem/mfem/commit/6aa01187d79e1e2b5f9a0f4819da3a1cc781cadd

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by v-dobrev** [comment]
There is a small issue with the following unit test: https://github.com/mfem/mfem/blob/de4a24b0ff7e723dfbdd7f6c0920ed372b98995b/tests/unit/linalg/test_cg_indefinite.cpp#L42-L48

Specifically, for this matrix and this initial guess, CG converges in one iteration, however, since we check for convergence with a strict inequality: https://github.com/mfem/mfem/blob/de4a24b0ff7e723dfbdd7f6c0920ed372b98995b/linalg/solvers.cpp#L648

we fail to detect convergence -- in this case the check is `0 < 0` since both tolerances are 0. If we make either of the CG tolerances (relative or absolute) positive, then the convergence is reported correctly.

I think we should switch the convergence test to non-strict inequality (`<=`) and come up with a different matrix or initial guess for the unit test.

**Comment by tzanio** [comment]
> I think we should switch the convergence test to non-strict inequality (<=) and come up with a different matrix or initial guess for the unit test.

I agree

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: CHANGELOG [hunk 1]
```diff
@@ -5,294 +5,255 @@
...
+  level builds on top of the Element Assembly kernels to compute a global sparse
+  matrix. All integrators supported by element assembly are also supported by
+  full assembly. See the '-fa' option in Example 9.
+
+- Optimized the AMD/HIP kernel support and enabled HIP support in the libCEED
+  integration. This is now available via the "ceed-hip" device backend.
+
+- Improved the libCEED integration to support:
+  * AssemblyLevel::NONE for Mass, Diffusion, VectorMass, and VectorDiffusion
+    Integrators. This level computes the full operator evalutaion "on the fly".
```
Comment (barker29) [inline CHANGELOG:41]:
```suggestion
    Integrators. This level computes the full operator evaluation "on the fly".
```

File: CHANGELOG [hunk 2]
```diff
@@ -5,294 +5,255 @@
...
-- Added a new miniapps/adjoint directory with two miniapps demonstrating how to
+- Added 10 new example codes:
+  * Example 25/25p demonstrates the use of a Perfectly Matched Layer (PML) for
+    electromagnetic wave propagation (indefinite Maxwell).
+  * Example 26/26p shows how to construct matrix-free geometric and p-multigrid
+    preconditioner for the Laplace problem.
+  * Example 27/27p demonstrates the enforcement of Dirichlet, Neumann, Robin,
+    and periodic boundary conditions with either H1 or DG Laplace problems.
+  * Versions of Example 1/1p in examples/amgx demonstrating the use of AmgX,
+    to solve th Laplace problem with AMG preconditioning on GPUs.
```
Comment (barker29) [inline CHANGELOG:172]:
```suggestion
    to solve the Laplace problem with AMG preconditioning on GPUs.
```

File: CHANGELOG [hunk 3]
```diff
@@ -5,294 +5,255 @@
...
+  * A version of Example 11p in examples/petcs demonstrating the use of SLEPc,
```
Comment (barker29) [inline CHANGELOG:173]:
```suggestion
  * A version of Example 11p in examples/petsc demonstrating the use of SLEPc,
```

**Review by barker29 — COMMENTED** [review]
Minor typos.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
This branch is now a release candidate for the `mfem-4.2` release.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Pull request merged**